File: scripts/lib/blog-comments.ts
export interface BlogComment {
  readonly author: string;
  readonly body: string;
  readonly createdAt: string;
  readonly isPriority: boolean;
}

const GISCUS_REPO = "bagrounds/obsidian-github-publisher-sync";

interface GqlDiscussionComment {
  readonly body: string;
  readonly author: { readonly login: string } | null;
  readonly createdAt: string;
}

interface GqlDiscussion {
  readonly title: string;
  readonly comments: { readonly nodes: readonly GqlDiscussionComment[] };
}

interface GqlSearchResult {
  readonly data?: { readonly search: { readonly nodes: readonly GqlDiscussion[] } };
  readonly errors?: ReadonlyArray<{ readonly message: string }>;
}

const toComment = (priorityUser: string | undefined) => (c: GqlDiscussionComment): BlogComment => ({
  author: c.author?.login ?? "unknown",
  body: c.body,
  createdAt: c.createdAt.split("T")[0] as string,
  isPriority: priorityUser !== undefined && c.author?.login === priorityUser,
});

const searchDiscussions = async (
  token: string,
  searchQuery: string,
  maxResults: number,
  maxComments: number,
): Promise<readonly GqlDiscussion[]> => {
  const query = `query($searchQuery: String!) {
    search(type: DISCUSSION, query: $searchQuery, first: ${maxResults}) {
      nodes { ... on Discussion { title, comments(first: ${maxComments}) { nodes { body, author { login }, createdAt } } } }
    }
  }`;

  const response = await fetch("https://api.github.com/graphql", {
    method: "POST",
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables: { searchQuery } }),
  });

  if (!response.ok) {
    console.warn(JSON.stringify({ event: "graphql_error", status: response.status }));
    return [];
  }
  const result = (await response.json()) as GqlSearchResult;
  if (result.errors) {
    console.warn(JSON.stringify({ event: "graphql_errors", errors: result.errors.map((e) => e.message) }));
    return [];
  }
  return result.data?.search?.nodes ?? [];
};

export const fetchGiscusComments = async (
  pathname: string,
  priorityUser: string | undefined,
): Promise<BlogComment[]> => {
  const token = process.env.GITHUB_TOKEN;
  if (!token) {
    console.log(JSON.stringify({ event: "skip_giscus", reason: "no GITHUB_TOKEN" }));
    return [];
  }
  try {
    const discussions = await searchDiscussions(token, `repo:${GISCUS_REPO} in:title "${pathname}"`, 1, 100);
    return (discussions[0]?.comments?.nodes ?? []).map(toComment(priorityUser));
  } catch (error) {
    console.warn(JSON.stringify({ event: "giscus_error", message: error instanceof Error ? error.message : String(error) }));
    return [];
  }
};

export const fetchAllSeriesComments = async (
  seriesId: string,
  priorityUser: string | undefined,
): Promise<BlogComment[]> => {
  const token = process.env.GITHUB_TOKEN;
  if (!token) {
    console.log(JSON.stringify({ event: "skip_giscus", reason: "no GITHUB_TOKEN" }));
    return [];
  }
  try {
    const discussions = await searchDiscussions(token, `repo:${GISCUS_REPO} in:title "${seriesId}/"`, 50, 50);
    return discussions
      .flatMap((d) => (d.comments?.nodes ?? []).map(toComment(priorityUser)))
      .sort((a, b) => b.createdAt.localeCompare(a.createdAt));
  } catch (error) {
    console.warn(JSON.stringify({ event: "giscus_error", message: error instanceof Error ? error.message : String(error) }));
    return [];
  }
};

File: haskell/src/Automation/BlogComments.hs
{-# LANGUAGE OverloadedStrings #-}

module Automation.BlogComments
  ( BlogComment (..)
  , fetchGiscusComments
  , fetchAllSeriesComments
  ) where

import Automation.BlogComments.GraphQL
  ( GqlComment
  , GqlDiscussion
  , GqlResponse
  )
import qualified Automation.BlogComments.GraphQL as Gql
import Automation.Json
  ( (.=)
  , encode
  , eitherDecode
  , object
  )
import Data.List (sortOn)
import Data.Ord (Down (..))
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import Network.HTTP.Client
  ( Manager
  , Request (..)
  , RequestBody (..)
  , httpLbs
  , parseRequest
  , responseBody
  , responseStatus
  )
import Network.HTTP.Types.Status (statusCode)
import System.Environment (lookupEnv)

giscusRepo :: Text
giscusRepo = "bagrounds/obsidian-github-publisher-sync"

graphqlEndpoint :: String
graphqlEndpoint = "https://api.github.com/graphql"

data BlogComment = BlogComment
  { author     :: Text
  , body       :: Text
  , createdAt  :: Text
  , isPriority :: Bool
  } deriving (Show, Eq)

toComment :: Maybe Text -> GqlComment -> BlogComment
toComment priorityUser graphqlComment = BlogComment
  { author     = maybe "unknown" Gql.login (Gql.author graphqlComment)
  , body       = Gql.body graphqlComment
  , createdAt  = Gql.createdAt graphqlComment
  , isPriority = case priorityUser of
      Just pu -> maybe False (\a -> Gql.login a == pu) (Gql.author graphqlComment)
      Nothing -> False
  }

buildQuery :: Int -> Int -> Text
buildQuery maxResults maxComments =
  "query($searchQuery: String!) { search(type: DISCUSSION, query: $searchQuery, first: "
    <> T.pack (show maxResults)
    <> ") { nodes { ... on Discussion { title, comments(first: "
    <> T.pack (show maxComments)
    <> ") { nodes { body, author { login }, createdAt } } } } } }"

searchDiscussions :: Manager -> Text -> Text -> Int -> Int -> IO [GqlDiscussion]
searchDiscussions manager token searchQuery maxResults maxComments = do
  initialRequest <- parseRequest graphqlEndpoint
  let requestBody' = encode $ object
        [ "query" .= buildQuery maxResults maxComments
        , "variables" .= object [ "searchQuery" .= searchQuery ]
        ]
  let httpReq = initialRequest
        { method = "POST"
        , requestBody = RequestBodyLBS requestBody'
        , requestHeaders =
            [ ("Authorization", "Bearer " <> TE.encodeUtf8 token)
            , ("Content-Type", "application/json")
            , ("User-Agent", "obsidian-github-publisher-sync")
            ]
        }
  response <- httpLbs httpReq manager
  let status = statusCode $ responseStatus response
  case status of
    200 ->
      case eitherDecode (responseBody response) of
        Left err -> do
          putStrLn $ "GraphQL parse error: " <> err
          pure []
        Right gqlResp ->
          case Gql.errors (gqlResp :: GqlResponse) of
            Just errs -> do
              putStrLn $ "GraphQL errors: " <> show (fmap Gql.message errs)
              pure []
            Nothing ->
              pure $ maybe [] (Gql.searchNodes . Gql.search) (Gql.responseData gqlResp)
    _ -> do
      putStrLn $ "GraphQL HTTP error: " <> show status
      pure []

fetchGiscusComments :: Manager -> Text -> Maybe Text -> IO [BlogComment]
fetchGiscusComments manager pathname priorityUser = do
  mToken <- lookupEnv "GITHUB_TOKEN"
  case mToken of
    Nothing -> do
      putStrLn "skip_giscus: no GITHUB_TOKEN"
      pure []
    Just token -> do
      let searchQuery = "repo:" <> giscusRepo <> " in:title \"" <> pathname <> "\""
      discussions <- searchDiscussions manager (T.pack token) searchQuery 1 100
      pure $ case discussions of
        []    -> []
        (d:_) -> fmap (toComment priorityUser) (Gql.nodes $ Gql.comments d)

fetchAllSeriesComments :: Manager -> Text -> Maybe Text -> IO [BlogComment]
fetchAllSeriesComments manager seriesId priorityUser = do
  mToken <- lookupEnv "GITHUB_TOKEN"
  case mToken of
    Nothing -> do
      putStrLn "skip_giscus: no GITHUB_TOKEN"
      pure []
    Just token -> do
      let searchQuery = "repo:" <> giscusRepo <> " in:title \"" <> seriesId <> "/\""
      discussions <- searchDiscussions manager (T.pack token) searchQuery 50 50
      let allComments = concatMap (fmap (toComment priorityUser) . Gql.nodes . Gql.comments) discussions
      pure $ sortOn (Down . createdAt) allComments
